vmd-api/api/openapi.yaml

```diff
@@ -0,0 +1,336 @@
+openapi: 3.1.0
+info:
+  contact:
+    email: nikita.bazulin@unikie.com
+  description: This is Control Panel API based on the OpenAPI 3.1 specification of
+    the GHAF project
+  license:
+    name: Apache License 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  title: Virtual Machine Manager API of the GHAF project - OpenAPI 3.1
+  version: 1.0.1
+externalDocs:
+  description: Find out more about the GHAF project
+  url: https://github.com/tiiuae/ghaf
+servers:
+- description: The primary server
+  url: https://localhost/api/v1
+- description: The test server
+  url: https://localhost/api/test/v1
+tags:
+- description: Operations of Virtual Machines
+  name: vm
+paths:
+  /vm/list:
+    get:
+      description: Return a IDs list of all Virtual Machines
+      operationId: getVmList
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                items: {}
+          description: IDs list of all Virtual Machines
+        default:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Get VMs list
+      tags:
+      - vm
+  /vm/info/{id}:
+    get:
+      description: Returns information about the Virtual Machine by ID
+      operationId: getVmInfoById
+      parameters:
+      - description: An ID of the Virtual Machine that needs to be fetched
+        explode: false
+        in: path
+        name: id
+        required: true
+        schema: {}
+        style: simple
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/VirtualMachineInfo'
+          description: successful operation
+        "400":
+          description: Invalid Virtual Machine ID supplied
+        "404":
+          description: Virtual Machine ID not found
+        default:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Finds Virtual Machine info by ID
+      tags:
+      - vm
+  /vm/{action}/{id}:
+    put:
+      description: Make an action to a Virtual Machine by it's ID
+      operationId: vmAction
+      parameters:
+      - description: "Action, which will be applied to a Virtual Machine"
+        example: start
+        explode: false
+        in: path
+        name: action
+        required: true
+        schema:
+          enum:
+          - start
+          - stop
+          - pause
+        style: simple
+      - description: Virtual Machine's ID
+        example: VM#101
```

Comment:
The `VM#101` ID encoding is redundant as the `VM` is clear from the URI context, also having it with the `VM` prefix would require that the server would return the list of IDs with all prefixed (also redundant). Let's drop the `VM`-prefix and only use the IDs. As with numeric unique IDs and we don't yet know how we generate the unique VM IDs on the server side so let's leave this example out for now.

Reply:
Should we use https://docs.rs/uuid/latest/uuid/?

Reply:
`uuid` (any) adds a mapping layer between `uuid` and `pid` (process id). The question is if we want that and what are the benefits the `uuid` brings (other than hide `pid` from client)? It will also require definitions like will `uuid` for a VM be persistent over restart (stop-start) cycles etc. 
With `pid` only, Linux will manage the IDs.

Reply:
No. And implementing superficial persistence Ida in server goes against representational state transfer statelessness design https://en.wikipedia.org/wiki/Representational_state_transfer#Statelessness

Reply:
KVM is a kernel virtualization module and I could not find `uuid`s that KVM would give to userland. It's basically so that a VMM (process) runs virtual machines using `ioctl`s to KVM API. What is seen on the userland side is process ID of the VMM process that is running the virtual machine.
If there's such uuid somewhere, I'd be happy to be proven wrong but please refer to documentation for exactly where this comes from.

Reply:
When creating a VM via KVM API on Linux, we have fd, file descriptor.
![screencapture](https://user-images.githubusercontent.com/1027150/226931829-dc2fc1fa-513d-43dc-9d55-a24346dd021f.png)
The VMM process (PID) can have a set of file descriptors under '/proc/<PID>/fd/` as the VM may access other resources, via their fds, as well.
Referring to VM fd (as u32) makes little sense as it's meant as the VMM (e.g. `crosvm`) internal. See more info on how KVM uses fds - https://www.kernel.org/doc/html/latest/virt/kvm/api.html#file-descriptors

Reply:
I wasn't referring to the fd, but the PID which is a u32.

But kvm documentation states that a vm is only associated with the fd and not with the process (PID).

_It is important to note that although VM ioctls may only be issued from the process that created the VM, a VM’s lifecycle is associated with its file descriptor, not its creator (process). In other words, the VM and its resources, including the associated address space, are not freed until the last reference to the VM’s file descriptor has been released. For example, if fork() is issued after ioctl(KVM_CREATE_VM), the VM will not be freed until both the parent (original) process and its child have put their references to the VM’s file descriptor._

[If we run a program like `virsh` we get fields ID and Name.](https://www.cyberciti.biz/faq/linux-list-a-kvm-vm-guest-using-virsh-command/)

Reading further into it [RedHat documentation talks about a Domain ID](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_deployment_and_administration_guide/sect-managing_guest_virtual_machines_with_virsh-retrieving_guest_virtual_machine_information) but it's unclear if this is a mapping done by the `virsh` program is this ID somehow associated with the file descriptor.

Furthermore KVM states about forking or dupping child processes:

_Because a VM’s resources are not freed until the last reference to its file descriptor is released, creating additional references to a VM via fork(), dup(), etc… without careful consideration is strongly discouraged and may have unwanted side effects, e.g. memory allocated by and on behalf of the VM’s process may not be freed/unaccounted when the VM is shut down._

So what I gather is that we definitely don't use information of the process such as the PID for identification, but using the file descriptor is possibly problematic if we spawn child processes.

The `virsh` program has the following identifiers for a vm:

- Domain name
- Domain ID
- UUID

Example:

```
# virsh dominfo guest1
Id:             8
Name:           guest1
UUID:           90e0d63e-d5c1-4735-91f6-20a32ca22c48
OS Type:        hvm
State:          running
CPU(s):         1
CPU time:       32.6s
Max memory:     1048576 KiB
Used memory:    1048576 KiB
Persistent:     yes
Autostart:      disable
Managed save:   no
Security model: selinux
Security DOI:   0
Security label: system_u:system_r:svirt_t:s0:c552,c818 (enforcing)
```

Reply:
Ok. I've been arguing for PID as it's independent of VMM (QEMU, rust-vmm-based VMMs, ...). However, as you introduced the virsh example, there are alternatives to `vmd` on managing guest VMs, often designed for cloud, rather than edge, virtualization management.

Please that ghaf uses microvm.nix generated systemd units. This is important distinction to virsh (and others).

See:
https://github.com/tiiuae/ghaf/blob/main/microvmConfigurations/netvm/default.nix
https://astro.github.io/microvm.nix/microvm-command.html#enabling-microvm-autostart
https://astro.github.io/microvm.nix/host-systemd.html

Target VMM is rust-vmm-based `crosvm`. In some cases we can use QEMU or other VMMs to develop and test concepts.

In practice, this means what I tried to describe earlier:
1. fd is `crosvm` internal. We can't use the fd to issue ioctls outside `crosvm` ("strongly discouraged")
2. We can analyze the `crosvm` process further based on `PID` - e.g. via `procfs`. This is valid, allowed and has no issues nor side effects.
3. The only case where we would use KVM API ioctls is to resolve the virtualization capabilities of the underlying HW when the `vmd` is started. This proved to be poorly supported by different HW so we may abandon direct dependency to KVM API and introduce vmd server internal dependency that better suits this purpose. Another key reason is that "no network on host requirement" allows vmd server to do only indirect queries to host (e.g. via shared memory).
So, we are not planning to use KVM API assigned `fd` of VM directly - that will be owned by the VMM on host.

In other words - analyzing running VMs using tools alternative to microvm.nix+VMM (like virsh) via fd is risky even if it would work (as you quoted the KVM documentation).

Given that, virsh concepts like "Domain name", "Domain ID" and "UUID" are irrelevant within this design context _unless_ you propose explicitly:
1. why those concepts are must in our context
2. what is the exact mechanism to generate those given our dependencies and constraints described above

Reply:
Study the usage of `pidfd` to implement VM ID via API